### PR TITLE
[#674, #675] Mitigate performance regression by filtering restricted registry collections

### PR DESCRIPTION
**Pull request by pavel-lexyr**
#675 introduces a significant performance regression in restricted registry collecting. This attempts to reduce the regression by re-introducing filter code removed in the previous version.

### CODE REVIEW

**Comment by pavel-lexyr** [comment]
py2.7 passes locally. What do the logs say on the CI?

**Comment by csmarchbanks** [comment]
```
==================================== ERRORS ====================================
_____________________ ERROR collecting tests/test_core.py ______________________
ImportError while importing test module '/home/circleci/project/tests/test_core.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/test_core.py:5: in <module>
    from unittest.mock import MagicMock
E   ImportError: No module named mock
```

You can use a skipIf and import only for the test such as done in https://github.com/prometheus/client_python/blob/master/tests/openmetrics/test_parser.py#L492.

**Comment by pavel-lexyr** [comment]
Aha. That worked - thank you!

**Review by csmarchbanks — APPROVED** [review]
This approach looks good to me, and I appreciate the test you added. I will give some time for @brian-brazil to review as well.

File: prometheus_client/registry.py [hunk 1]
```diff
@@ -135,10 +135,20 @@ def __init__(self, names, registry):
         self._registry = registry
 
     def collect(self):
-        for metric in self._registry.collect():
-            m = metric._restricted_metric(self._name_set)
-            if m:
-                yield m
+        names = copy.copy(self._name_set)
```
Comment (brian-brazil) [inline prometheus_client/registry.py:138]: Why is this copy here? This set shouldn't be changing after the object is instantiated.

File: prometheus_client/registry.py [hunk 2]
```diff
@@ -135,10 +135,20 @@ def __init__(self, names, registry):
...
+        collectors = set()
+        with self._registry._lock:
+            if 'target_info' in names and self._registry._target_info:
+                yield self._registry._target_info_metric()
```
Comment (brian-brazil) [inline prometheus_client/registry.py:142]: I don't think it is safe to yield under the lock.

File: prometheus_client/registry.py (hunk 1 above)
Comment (pavel-lexyr) [reply prometheus_client/registry.py:138]: Trying to do it in the same way collections used to be filtered with target_info. There used to be a set variable that was modified if target_info was found - in order not to change the class field in the same way, we copy it to `names`.

File: prometheus_client/registry.py (hunk 2 above)
Comment (pavel-lexyr) [reply prometheus_client/registry.py]: Fixed. Good catch!

File: prometheus_client/registry.py [hunk 3]
```diff
@@ -135,10 +135,23 @@ def __init__(self, names, registry):
...
-            m = metric._restricted_metric(self._name_set)
-            if m:
-                yield m
+        names = copy.copy(self._name_set)
+        collectors = set()
+        yield_target_info = False
+        with self._registry._lock:
+            if 'target_info' in names and self._registry._target_info:
+                yield_target_info = True
+                names.remove('target_info')
```
Comment (brian-brazil) [inline prometheus_client/registry.py:144]: You don't need this, the registry should be preventing this from happening in the first place.

File: prometheus_client/registry.py [hunk 4]
```diff
@@ -135,10 +135,23 @@ def __init__(self, names, registry):
         self._registry = registry
 
     def collect(self):
-        for metric in self._registry.collect():
-            m = metric._restricted_metric(self._name_set)
-            if m:
-                yield m
+        names = copy.copy(self._name_set)
+        collectors = set()
+        yield_target_info = False
```
Comment (brian-brazil) [inline prometheus_client/registry.py:140]: I'd have this as the thing to yield, rather than a bool for cleanliness. Also avoids a race.

Comment (pavel-lexyr) [reply prometheus_client/registry.py]: Done.

File: prometheus_client/registry.py (hunk 3 above)
Comment (pavel-lexyr) [reply prometheus_client/registry.py]: Not sure I understood it right - but done.

**Comment by brian-brazil** [comment]
:+1:

**Review by csmarchbanks — APPROVED** [review]
Thanks for the additional review and fixes!

**Pull request merged**